First version

LoyVanBeek opened the pull request:
Merge together/after: https://github.com/tue-robotics/cb_base_navigation_msgs/pull/7

LoyVanBeek committed:
Add basic ROS package files

LoyVanBeek committed:
Add messages and services

LoyVanBeek committed:
Make the package compile at least

LoyVanBeek committed:
Renme EdgeTraversal to just Edge and add a service to update an edge

LoyVanBeek committed:
Rename nodes of an Edge to just u,v to match literature

LoyVanBeek committed:
WIP: Starting to set up basic versions of topo planning services

LoyVanBeek committed:
fixup! Renme EdgeTraversal to just Edge and add a service to update an edge

LoyVanBeek committed:
Implement mocked get_plan service

LoyVanBeek committed:
Smarter mock: insert a door between the query origin and destination

LoyVanBeek committed:
WIP: set up a dummy graph but planning in it kinda works

LoyVanBeek committed:
Rename u,v to origin, destination for clarity

jlunenburg committed:
fix(tap)Placed services in node namespace

LoyVanBeek committed:
Instead of bool success, use an error message & code

LoyVanBeek committed:
Use strings in a Enum for the action type

MatthijsBurgh committed:
Convert to python library

MatthijsBurgh committed:
Split msgs from library

LoyVanBeek committed:
Add example graph definition, which can be loaded with

rosparam load tests/example_graph.yaml /topological_action_planner/edges
and then queried with rosservice call /topological_action_planner/getPlan ...

MatthijsBurgh committed:
Remove unused msg/srv defs

LarsJanssenTUe committed:
Initial commit marker publisher

LoyVanBeek committed:
Fix typos in message defintions

LoyVanBeek committed:
Move some serialisation and deserialisation to a separate file

LoyVanBeek committed:
Move visualisation functions to util

LoyVanBeek committed:
Fix typo

LarsJanssenTUe committed:
Working prototype markerarray

LarsJanssenTUe committed:
Dummy for Loy

LoyVanBeek committed:
Clean up old imports

LoyVanBeek committed:
Publish (dummy) visualisation of the graph

LoyVanBeek committed:
fixup! Publish (dummy) visualisation of the graph

LoyVanBeek committed:
Make plotting optional, disabled by default

LoyVanBeek committed:
Implement service to update edges

LoyVanBeek committed:
Use weight of edges for planning and log path length

jlunenburg committed:
feat(launch)Added launch file

jlunenburg committed:
test(graph)Added stuff to example graph

MatthijsBurgh committed:
fix newline at EOF

LarsJanssenTUe committed:
Untested visualisation

LoyVanBeek committed:
WIP: busy setting up integration with ED

LoyVanBeek committed:
Set up basic, mocked ED interface

LoyVanBeek committed:
Iterate to find the lowest cost plan

LoyVanBeek committed:
Configure cost per action

LoyVanBeek committed:
Visualization fixups and setting up a dummy graph for Biestheuvel env

LoyVanBeek committed:
Fixup syntax to assign updated weights AND stub out global_planner call temporarily

LoyVanBeek committed:
Fixup node insertion for the current robot pose

LoyVanBeek committed:
Determine room for both entiy and area

LoyVanBeek committed:
Call global_planner/get_plan_srv to get length of a path

LoyVanBeek committed:
Calculate the cost of traveling based on distance

LoyVanBeek committed:
Update env mocked knowledge a bit for more interesting scenario

LoyVanBeek committed:
Squelch annoying messages about uninitialized quats in rviz

LoyVanBeek committed:
Busy integrating with ed_python. WIP

MatthijsBurgh committed:
(actions) add GH action

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/serialisation.py:
```diff
@@ -0,0 +1,44 @@
+from typing import Sequence, Mapping
+
+import genpy
```
Missing dependency in package.xml

MatthijsBurgh commented on topological_action_planner_msgs/msg/Node.msg:
```diff
@@ -0,0 +1,2 @@
+string entity
+string area
```
Missing newline at EOF

MatthijsBurgh commented on topological_action_planner_msgs/srv/UpdateEdge.srv:
```diff
@@ -0,0 +1,3 @@
+Edge updated
+---
+bool success
```
Missing newline at EOF

MatthijsBurgh commented on topological_action_planner/tests/example_graph.yaml:
```diff
@@ -0,0 +1,25 @@
+  destination:
+    entity: 'coffee_table'
+    area: in_front_of
+
```
Remove empty line

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/util.py:
```diff
@@ -0,0 +1,36 @@
+
+
+def visualize(graph):
+    import matplotlib.pyplot as plt
```
Missing dep on matplotlib in package.xml

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/ed_interface.py:
```diff
@@ -0,0 +1,61 @@
+from __future__ import print_function
```
Not needed as we dropped python2 support.

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/ed_interface.py:
```diff
@@ -0,0 +1,61 @@
+from __future__ import print_function
+
+# TU/e
+import rospy
```
This file is going to be replaced with ed_python

LoyVanBeek replied:
Hope so. That work is on a  different branch though for now.

MatthijsBurgh commented on topological_action_planner/scripts/tap_marker_publisher.py:
```diff
@@ -0,0 +1,39 @@
+    try:
+        pub_marker()
+    except rospy.ROSInterruptException:
+        pass
```
Missing newline EOF

MatthijsBurgh reviewed (commented):
We should black all new python code IMO

MatthijsBurgh committed:
Convert msgs deps to exec_depend

MatthijsBurgh committed:
Add dep to ed_python

MatthijsBurgh committed:
Fill __init__.py

MatthijsBurgh committed:
Add missing empty line

MatthijsBurgh committed:
Remove breakpoint

MatthijsBurgh committed:
Small fixes in visualisation

LarsJanssenTUe committed:
Update(visualisation): Different edge types have different colors

LoyVanBeek committed:
Address a bunch of formatting review ceomments

LoyVanBeek committed:
black-ed all the code

Re-blacked all the code with 120 char lines

MatthijsBurgh committed:
(actions) add black linting

MatthijsBurgh committed:
Create README.md

LoyVanBeek committed:
Deal with entities not exisiting and hndle KDL datatypes correctly

LoyVanBeek committed:
Remove ed_interface but use ed_python instead

LoyVanBeek committed:
No need to get entity 2x

MatthijsBurgh committed:
Add docs

LoyVanBeek committed:
When looking up rooms_of_volume, get  the entity first

LoyVanBeek committed:
Get the current room the robot is in, kinda

LoyVanBeek committed:
Merge branch 'biestheuvel_ed_python_integration' into biestheuvel

MatthijsBurgh committed:
Merge remote-tracking branch 'origin/biestheuvel_ed_python_integration' into biestheuvel

MatthijsBurgh committed:
Fix linting

MatthijsBurgh committed:
Fix linting

LoyVanBeek committed:
Get constraint area correctly

LoyVanBeek committed:
Get constraint area correctly

MatthijsBurgh committed:
Remove setup.py from msgs

MatthijsBurgh committed:
Fix linting

MatthijsBurgh commented on topological_action_planner/scripts/tap.py:
```diff
@@ -0,0 +1,217 @@
+
+            self._pub_grasp_marker.publish(create_tap_marker_array(graph, self.wm))
+            rospy.loginfo("Found plan of {} edges".format(len(edges)))
+            return PlanResponse(error_msg="", error_code=PlanResponse.SUCCESS, edges=edges)
+        except nx.NetworkXNoPath as no_path_found_ex:
+            rospy.logerr(no_path_found_ex)
+            return PlanResponse(
+                error_msg=str(no_path_found_ex),
+                error_code=PlanResponse.ERROR_NO_PATH_FOUND,
+            )
+        except nx.NodeNotFound as node_not_found_ex:
+            rospy.logerr(node_not_found_ex)
+            return PlanResponse(
+                error_msg=str(node_not_found_ex),
+                error_code=PlanResponse.ERROR_UNKNOWN_NODE,
+            )
+
+    def _srv_update_edge_cb(self, req: UpdateEdgeRequest) -> UpdateEdgeResponse:
+        try:
+            self.update_edge(req.updated)
+            return UpdateEdgeResponse(success=True)
+        except nx.NodeNotFound as node_not_found_ex:
+            rospy.logerr(node_not_found_ex)
+            return UpdateEdgeResponse(success=False)
+
+    def update_edge(self, edge: Edge):
+        u, v = (edge.origin.entity, edge.origin.area), (
+            edge.destination.entity,
+            edge.destination.area,
+        )
+        self.G[u][v]["weight"] = edge.cost
+        self.G[u][v]["action_type"] = edge.action_type
+
+    def get_area_constraint(self, entity: str, area: str) -> PositionConstraint:
+        """What is the center pose of an entity and area?"""
+        res = self._get_constraint_srv(entity_ids=[entity], area_names=[area])
+        if not res.error_msg:
+            return PositionConstraint(constraint=res.position_constraint_map_frame, frame="map")
+        else:
+            raise Exception("Cannot get a constraint for {}.{}".format(entity, area))
```
Should be moved to library

LoyVanBeek committed:
Fixup visualisation

LoyVanBeek committed:
Adjust grasph to robotics_lab instead of Biestheuvel

LoyVanBeek committed:
Vizualize edges in two directions

LoyVanBeek committed:
Get robot name from system instead of hardcoding

LoyVanBeek committed:
Some error logging

LoyVanBeek committed:
Handle case where origin is the robot

LoyVanBeek committed:
Cleanup obsolete todo

LoyVanBeek committed:
Apply black

LarsJanssenTUe committed:
Remove unused and redundant code

LoyVanBeek committed:
Add a bunch of docstrings

MatthijsBurgh committed:
Add dependabot config for submodules

LoyVanBeek committed:
Apply black

LoyVanBeek committed:
Re-add .gitignore

LoyVanBeek committed:
Move core planning function to library

LoyVanBeek committed:
fixup! Move core planning function to library

LoyVanBeek committed:
Apply black -l 120

MatthijsBurgh commented on README.md:
```diff
@@ -1,2 +1,7 @@
+
+[![CI](https://github.com/tue-robotics/topological_action_planner/actions/workflows/main.yml/badge.svg)](https://github.com/tue-robotics/topological_action_planner/actions/workflows/main.yml) [![Lint](https://github.com/tue-robotics/topological_action_planner/actions/workflows/black.yml/badge.svg)](https://github.com/tue-robotics/topological_action_planner/actions/workflows/black.yml)
+
+ROS package for planning over a topological navigation mesh. 
```
Mesh?

MatthijsBurgh commented on README.md:
```diff
@@ -1,2 +1,7 @@
+
+ROS package for planning over a topological navigation mesh. 
+Nodes are areas of interest in an ED world model and edges between these areas are labeled with the action to take to traverse the edge,
+    like opening a door, plain driving or pushing an object aside.
```
```suggestion
like opening a door, plain driving or pushing an object aside.
```

MatthijsBurgh commented on topological_action_planner/README.md:
````diff
@@ -0,0 +1,39 @@
+Run 
+```bash
+roslaunch topological_action_planner start.launch
+rosservice call /hero/topological_action_planner/get_plan  "origin:
````
Remove reference to hero

LoyVanBeek replied:
No, I want something I can copy-paste here.

MatthijsBurgh replied:
I am  not sure, this is the place for that. Not the root readme IMHO.

MatthijsBurgh commented on topological_action_planner/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+  DESTINATION ${CATKIN_PACKAGE_BIN_DESTINATION}
+)
+
+## Add gtest based cpp test target and link libraries
+# catkin_add_gtest(${PROJECT_NAME}-test test/test_topological_action_planner.cpp)
+# if(TARGET ${PROJECT_NAME}-test)
+#   target_link_libraries(${PROJECT_NAME}-test ${PROJECT_NAME})
+# endif()
```
Remove commented code

MatthijsBurgh commented on topological_action_planner/CMakeLists.txt:
```diff
@@ -0,0 +1,29 @@
+#   target_link_libraries(${PROJECT_NAME}-test ${PROJECT_NAME})
+# endif()
+
+# catkin_add_nosetests(test)
```
We should add tests

MatthijsBurgh commented on topological_action_planner/README.md:
````diff
@@ -0,0 +1,39 @@
+```
+or 
+```bash
+rosservice call /hero/topological_action_planner/get_plan  "origin:
````
Again, remove reference to hero

LoyVanBeek replied:
Again, I want to be able to copy-paste for testing this easily. I'll add a note though

MatthijsBurgh commented on README.md:
```diff
@@ -1,2 +1,7 @@
 # topological_action_planner
```
Run through MD linter, https://dlaa.me/markdownlint/

MatthijsBurgh commented on topological_action_planner/README.md:
```diff
@@ -0,0 +1,39 @@
+# topological_action_planner
```
Run through MD linter, https://dlaa.me/markdownlint/

MatthijsBurgh commented on topological_action_planner/scripts/tap.py:
```diff
@@ -0,0 +1,149 @@
+        self._srv_update_edge = rospy.Service("~update_edge", UpdateEdge, self._srv_update_edge_cb)
+        self._pub_grasp_marker = rospy.Publisher("~graph", MarkerArray, queue_size=10, latch=True)
+
+        # self.G = generate_dummy_graph()
+        # to_yaml(self.G, '/tmp/graph.yaml')
```
Remove commented code

MatthijsBurgh commented on topological_action_planner/scripts/tap.py:
```diff
@@ -0,0 +1,149 @@
+            node0_entity = self._wm.get_entity(node[0])
+            graph.nodes[node]["room"] = rooms_of_volume(
+                self._wm, node0_entity, node[1]
+            )  # Based on where they are in ED?
```
Comment is unclear to me

MatthijsBurgh commented on topological_action_planner/scripts/tap.py:
```diff
@@ -0,0 +1,149 @@
+            )  # Based on where they are in ED?
+
+        if req.origin.entity == "":
+            # This indicates the plan starts from the robot's current pose
```
Should we document these design choices somewhere?

LoyVanBeek replied:
It's now in the .srv definition

MatthijsBurgh commented on topological_action_planner/scripts/tap.py:
```diff
@@ -0,0 +1,149 @@
+        try:
+            self.update_edge(req.updated)
+            return UpdateEdgeResponse(success=True)
+        except nx.NodeNotFound as node_not_found_ex:
+            rospy.logerr(node_not_found_ex)
```
```suggestion
        except nx.NodeNotFound as ex:
            rospy.logerr(ex)
```

LoyVanBeek replied:
Do not agree, there's multiple exception types, I want to be clear about their differences.

MatthijsBurgh replied:
That is clear from the `except XXX` imo.

MatthijsBurgh commented on topological_action_planner/scripts/tap.py:
```diff
@@ -0,0 +1,149 @@
+                error_msg=str(no_path_found_ex),
+                error_code=PlanResponse.ERROR_NO_PATH_FOUND,
+            )
+        except nx.NodeNotFound as node_not_found_ex:
+            rospy.logerr(node_not_found_ex)
+            return PlanResponse(
+                error_msg=str(node_not_found_ex),
```
```suggestion
        except nx.NodeNotFound as ex:
            rospy.logerr(ex)
            return PlanResponse(
                error_msg=str(ex),
```

LoyVanBeek replied:
Same

MatthijsBurgh commented on topological_action_planner/scripts/tap.py:
```diff
@@ -0,0 +1,149 @@
+            self._pub_grasp_marker.publish(create_tap_marker_array(graph, self._wm))
+            rospy.loginfo("Found plan of {} edges".format(len(edges)))
+            return PlanResponse(error_msg="", error_code=PlanResponse.SUCCESS, edges=edges)
+        except nx.NetworkXNoPath as no_path_found_ex:
+            rospy.logerr(no_path_found_ex)
+            return PlanResponse(
+                error_msg=str(no_path_found_ex),
```
```suggestion
        except nx.NetworkXNoPath as ex:
            rospy.logerr(ex)
            return PlanResponse(
                error_msg=str(ex),
```

LoyVanBeek replied:
Same

MatthijsBurgh commented on topological_action_planner/scripts/tap.py:
```diff
@@ -0,0 +1,149 @@
+            # This indicates the plan starts from the robot's current pose
+            # We'll insert a node corresponding to our current position.
+            # Difficulty is determining it's adjacency to other nodes, based on geometry
+            robot_name = rospy.get_namespace().split("/")[-2]
```
This should be done in the constructor. Also I am sure this isn't robust, as if run this in other namespaces, `-2` is not the first anymore.

LoyVanBeek replied:
I have no alternative AFAIK

MatthijsBurgh replied:
It is a list, so why not take the first item?

MatthijsBurgh commented on topological_action_planner/scripts/tap.py:
```diff
@@ -0,0 +1,149 @@
+            edge.destination.entity,
+            edge.destination.area,
+        )
+        self.G[u][v]["weight"] = edge.cost
+        self.G[u][v]["action_type"] = edge.action_type
```
I expect `IndexError`/`KeyError` very soon here

LoyVanBeek replied:
That kind of exception is handled by it's caller. It is OK to raise an exception here, accessing those items is the only thing the function does. If it is given wrong input, let the caller know.

MatthijsBurgh commented on topological_action_planner/scripts/tap.py:
```diff
@@ -0,0 +1,149 @@
+from cb_base_navigation_msgs.msg import PositionConstraint
+
+
+class TopoPlannerNode:
```
Move to library, not inside script folder

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/planning.py:
```diff
@@ -0,0 +1,116 @@
+
+    def plan(self, graph: nx.Graph, origin_node: Node, dst_node: Node) -> List[Edge]:
+        lowest_total_cost = float("inf")
+        while True:
```
I would prefer `not rospy.is_shutdown()`

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/planning.py:
```diff
@@ -0,0 +1,116 @@
+            for edge in edges:
+                if edge.action_type == Edge.ACTION_DRIVE:
+                    # TODO: we should actually plan from the end of the plan found for the previous edge.
+                    # Otherwise the center pose could be blocked but not e whole area and we would still fail.
```
Typo and not really clear to me

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/planning.py:
```diff
@@ -0,0 +1,116 @@
+                        # Cannot plan along this edge
+                        edge_cost = 100
+
+                    rospy.loginfo(
```
Use f-string

LoyVanBeek committed:
Split off functionality to update an edge's cost

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/util.py:
```diff
@@ -0,0 +1,46 @@
+    """
+    numbers = "0123456789"
+    letters = "abcdefghijklmnopqrstuvwxyz"
+    import random
```
Why not import at the top?

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/util.py:
```diff
@@ -0,0 +1,46 @@
+from topological_action_planner_msgs.msg import Edge
+
+
+def generate_dummy_graph(n_nodes=10, n_edges=20) -> nx.Graph:
```
So this is for testing only? Then it shouldn't be in the library

LoyVanBeek replied:
Where then?

MatthijsBurgh replied:
In a test/example folder

LoyVanBeek replied:
It's not an example, it is a utility function that is currently unused. It might be useful in a unit test though

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/visualisation.py:
```diff
@@ -0,0 +1,124 @@
+        e_mark.pose.orientation.w = 1  # To squelch annoying messages about uninitialized quats
+        origin_entity = wm.get_entity(origin[0])
+        if not origin_entity:
+            rospy.logwarn("No such entity '{}'".format(origin[0]))
```
Please use f-string

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/visualisation.py:
```diff
@@ -0,0 +1,124 @@
+            continue
+
+        if origin[1]:
+            vec1 = wm.tf_buffer.transform(
+                VectorStamped(
+                    origin_entity.volumes[origin[1]].center_point, frame_id=origin_entity.uuid, stamp=rospy.Time(0)
+                ),
+                "map",
+            )
+        else:
+            vec1 = VectorStamped.from_framestamped(wm.tf_buffer.transform(origin_entity.pose, "map"))
```
I don't like that we switch the order of creating the VectorStamped and transforming it.
Better to split over multiple lines anyway

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/visualisation.py:
```diff
@@ -0,0 +1,124 @@
+
+        destination_entity = wm.get_entity(destination[0])
+        if not destination_entity:
+            rospy.logwarn("No such entity '{}'".format(destination[0]))
```
f-string

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/visualisation.py:
```diff
@@ -0,0 +1,124 @@
+            continue
+
+        if destination[1]:
+            vec2 = wm.tf_buffer.transform(
+                VectorStamped(
+                    destination_entity.volumes[destination[1]].center_point,
+                    frame_id=destination_entity.uuid,
+                    stamp=rospy.Time(0),
+                ),
+                "map",
+            )
+        else:
+            vec2 = VectorStamped.from_framestamped(wm.tf_buffer.transform(destination_entity.pose, "map"))
```
Again order of transforming and class creation

Better to split over multiple lines anyway

MatthijsBurgh commented on topological_action_planner/src/topological_action_planner/visualisation.py:
```diff
@@ -0,0 +1,124 @@
+        reversed_arrow.points = list(reversed(reversed_arrow.points))
+        marker_array.markers.append(reversed_arrow)
+
+        rospy.loginfo("Add edge {} -> {}".format(origin, destination))
+        rospy.loginfo("{} -> {}".format(origin_entity, destination_entity))
+        rospy.loginfo("{} -> {}".format(vec1, vec2))
```
f-strings, please

MatthijsBurgh commented on topological_action_planner/tests/example_graph.yaml:
```diff
@@ -0,0 +1,32 @@
+- action_type: DRIVE
```
I think this should be in an example folder

MatthijsBurgh commented on topological_action_planner_msgs/README.md:
```diff
@@ -0,0 +1,2 @@
+# topological_action_planner
+ROS package for plannng over a topological navigation mesh. Nodes are areas of interest in an ED world model and edges between these areas are labeled with the action to take to traverse the edge, like opening a door or plain driving.
```
Change, as this is the msgs package.

MatthijsBurgh commented on topological_action_planner_msgs/msg/Edge.msg:
```diff
@@ -0,0 +1,8 @@
+Node origin
+Node destination
+string ACTION_DRIVE=DRIVE
+string ACTION_OPEN_DOOR=OPEN_DOOR
+string ACTION_PUSH_OBJECT=PUSH_OBJECT
+#string ACTION_USE_ELEVATOR=USE_ELEVATOR
+string action_type  # Any of ACTION_*
+float32 cost  # How expensive is is to travel along this edge
```
Maybe add some documentation, at least some whitespace

LoyVanBeek committed:
Split up TopoPlanner to use an dependency injected object to calculate edge costs

This allows to test the TopoPlanner without depending on Ed and a navigation planner

LoyVanBeek committed:
Add some basic tests for the TopoPlanner

LoyVanBeek committed:
Address review comments

LoyVanBeek committed:
Review: more f-strings

LoyVanBeek committed:
Some interface docs

LoyVanBeek committed:
Remove 'hero' from canonical service name but add note about it

LoyVanBeek committed:
Rename exception  objects to just 'ex' instead of type-specific name

LoyVanBeek committed:
Move node code to library as well

LoyVanBeek committed:
Shuffle lines for prettyness

LoyVanBeek committed:
Move example_graph to example dir

LoyVanBeek committed:
Apply black -l 120, mostly on tests

LoyVanBeek committed:
Refactor test to instantiate planner in test-method itself

LoyVanBeek committed:
Add test for effect of EdgeCostCalculator

MatthijsBurgh committed:
remove unneeded find_package of rostest

MatthijsBurgh committed:
Remove unused import

MatthijsBurgh committed:
Remove duplicate code from test

MatthijsBurgh committed:
Add import test

MatthijsBurgh committed:
Fix linting of tests

MatthijsBurgh committed:
Fix imports

MatthijsBurgh commented on topological_action_planner_msgs/msg/Edge.msg:
```diff
@@ -0,0 +1,11 @@
+string ACTION_DRIVE=DRIVE
+string ACTION_OPEN_DOOR=OPEN_DOOR
+string ACTION_PUSH_OBJECT=PUSH_OBJECT
+#string ACTION_USE_ELEVATOR=USE_ELEVATOR
```
Why commented, I think we can prepare for feature functions

LoyVanBeek committed:
Include USE_ELEVATOR as action_type

LoyVanBeek commented:
Good to merge, @MatthijsBurgh

MatthijsBurgh commented:
Needs to be merged together with https://github.com/tue-robotics/cb_base_navigation_msgs/pull/7, which requires other PRs to be merged, but I will approve this PR

MatthijsBurgh committed:
ed_py renamed to ed
